docs/spec/overload.rst

```diff
@@ -99,8 +109,397 @@ variable is not sufficient.
+argument list. Unlike step 1, this step considers the types of the parameters
+and arguments. During this step, do not generate any user-visible errors.
+Simply record which of the overloads result in evaluation errors.
+
+- If all overloads result in errors, proceed to step 3.
+- If only one overload evaluates without error, it is the winning match.
+  Evaluate it as if it were a non-overloaded function call and stop.
+- If two or more candidate overloads remain, proceed to step 4.
+
+
+Step 3: If step 2 produces errors for all overloads, perform
+"argument type expansion". Some types can be decomposed
+into two or more subtypes. For example, the type ``int | str`` can be
+expanded into ``int`` and ``str``.
+
+Expansion should be performed one argument at a time from left to
+right. Each expansion results in sets of effective argument types.
+For example, if there are two arguments whose types evaluate to
+``int | str`` and ``int | bytes``, expanding the first argument type
+results in two sets of argument types: ``(int, ?)`` and ``(str, ?)``.
+Here ``?`` represents an unexpanded argument type.
+If expansion of the second argument is required, four sets of
+argument types are produced: ``(int, int)``, ``(int, bytes)``,
+``(str, int)``, and ``(str, bytes)``.
+
+After each argument expansion, return to step 2 and evaluate all
+expanded argument lists.
+
+- If all argument lists evaluate successfully, combine their
+  respective return types by union to determine the final return type
+  for the call, and stop.
+- If argument expansion has been applied to all arguments and one or
+  more of the expanded argument lists cannot be evaluated successfully,
+  generate an error and stop.
+
+
+For additional details about argument type expansion, see
+`argument-type-expansion`_ below.
+
+
+Step 4: If the argument list is compatible with two or more overloads,
+determine whether one or more of the overloads has a variadic parameter
+(either ``*args`` or ``**kwargs``) that maps to a corresponding argument
+that supplies an indeterminate number of positional or keyword arguments.
+If so, eliminate overloads that do not have a variadic parameter.
+
+- If this results in only one remaining candidate overload, it is
+  the winning match. Evaluate it as if it were a non-overloaded function
+  call and stop.
+- If two or more candidate overloads remain, proceed to step 5.
+
+
+Step 5: If the type of one or more arguments evaluates to a
+type that includes a :term:`gradual form` (e.g. ``list[Any]`` or
+``str | Any``), determine whether some theoretical
+:term:`materialization <materialize>` of these gradual types could be used
+to disambiguate between two or more of the remaining overloads.
+
+- If none of the arguments evaluate to a gradual type, proceed to step 6.
+- If one or more arguments evaluate to a gradual type but no possible
+  materializations of these types would disambiguate between the remaining
+  overloads, proceed to step 6.
```

Comment:
These two bullet points still use the term "gradual type" in an imprecise way, since all types are gradual types.

Since all types are gradual types, and all types can be materialized (though the only possible materialization of a fully static type is itself), I think this step would be simpler if it eliminated all conditionality based on the original argument type, and just discussed materialization directly. E.g. change the first paragraph to:

> Step 5: determine whether any possible materialization of any argument type could disambiguate between two or more of the remaining overloads.

And collapse these two bullets to one:

```suggestion
- If no possible materialization of any argument type would disambiguate between the remaining
  overloads, proceed to step 6.
```

Reply:
I took another swing at redefining this rule, and I'm much happier with the latest version. It's somewhat different (and more precise) than what you are proposing above. It's also subtly different from what mypy and pyright implement today, but I think it's more aligned with the concept of type manifestations and the "gradual guarantee".